File: dart/lib/src/router.dart

import 'dart:math';

import 'context.dart';

typedef Handler = Future<void> Function(Context);

class Router {
  String uri;
  String method;

  List<Handler> handlers;

  Map<String, Router> _children;
  Map<String, Router> _regexpChildren;
  Map<String, Router> _universalChildren;

  Router(String method, String uri) {
    this.method = method;
    if (uri.indexOf("/") == 0) {
      this.uri = uri.substring(1);
    } else {
      this.uri = uri;
    }
    this.handlers = [];
    this._children = new Map<String, Router>();
    this._regexpChildren = new Map<String, Router>();
    this._universalChildren = new Map<String, Router>();
  }

  Router _add(String method, String uri, Handler fn) {
    List<String> uris = uri.split("/").sublist(1);
    Router r = this;
    for (int i = 0; i < uris.length; i++) {
      print(uris[i]);
      Router lr = new Router(method, uris[i]);
      if (lr.uri.indexOf(":") == 0) {
        // 通配
        lr.uri = uris[i].substring(1);
        r._universalChildren[method] = lr;
      } else if (lr.uri.indexOf("#") == 0) {
        // 正则
        lr.uri = uris[i].substring(1);
        r._regexpChildren[method] = lr;
      } else {
        // 精确
        r._children[method+"-"+uris[i]] = lr;
      }
      r = lr;
    }
    if (fn != null) r.handlers = [fn];
    return r;
  }

  Router use(Handler fn) {
    this.handlers.add(fn);
    return this;
  }

  Router hook(Router r) {
    this._children["${r.method}-${r.uri}"] = r;
    return this;
  }

  get(String uri, Handler fn) {
    this._add("GET", uri, fn);
  }

  post(String uri, Handler fn) {
    this._add("POST", uri, fn);
  }

  Router group(String uri) {
    return this._add("ANY", uri, null);
  }

  find(Context ctx) {
    List<String> uris = ctx.uri.split("/").sublist(1);
    uris.add("");
    Router r = this;
    ctx.handlers.addAll(r.handlers);
    for(var i = 0; i < uris.length; i++) {
      Router nr = r._children["${ctx.method}-${uris[i]}"] ?? r._children["ANY-${uris[i]}"];
      // 判断正则
      if (nr == null && r._regexpChildren[ctx.method] != null) {
        Router lr = r._regexpChildren[ctx.method];
        var reg = new RegExp(lr.uri);
        if (reg.hasMatch(uris[i])) {
          nr = lr;
          // 正则匹配逻辑
          reg.allMatches(uris[i]).forEach((v) {
            for(var g in v.groupNames.toList()) {
              ctx.params[g] = v.namedGroup(g);
            }
          });
        }
      }
      // 判断通配
      if (nr == null && r._universalChildren[ctx.method] != null) {
        nr = r._universalChildren[ctx.method];
        ctx.params[nr.uri] = uris[i];
      }
      // 增加404
      if (nr == null) {
        ctx.handlers.add((Context ctx) async {
          ctx.statusCode = 404;
          ctx.send("Not found");
        });
      } else {
        ctx.handlers.addAll(nr.handlers);
        r = nr;
      }
    }
    ctx.next();
  }
}

File: nodejs/src/web/router.js
/**
 * @name Router 路由相关处理
 */

class Router {
    /**
     * @name 子路由
     * @type {Map<Method, Router>}
     */
     childrenRoute = new Map();
    /**
     * @name 路由路径uri
     * @type {String}
     */
    uri;
    /**
     * @name 中间件
     * @type {Array<Handler>}
     */
    middleware = [];
    /**
     * @name 正则匹配路由
     * @type {Map<Method, Router>}
     */
    regexChildren = new Map();
    /**
     * @name 通配路由
     * @type {Map<Method, Router>}
     */
    universalChildren = new Map();
    /**
     * @name 注册方法
     * @type {String}
     */
    method;

    /**
     * @param {String} u 子路由uri
     */
    constructor(u) {
        u = u || '';
        if (u.indexOf('/') == 0) u = u.substr(1);
        if (u.indexOf(':') == 0) {
            this.type = 'universal';
        }

        if (u.indexOf('#') == 0) {
            this.type = 'regex';
        }

        this.uri = u;
    }

    hook(...routers) {
        for(let item of routers) {
            this.childrenRoute.set(`ANY-${item.uri}`, item);
        }
    }

    /**
     * @name 中间件
     * @param  {...Handler} handlers
     */
    use(...handlers) {
        this.middleware.push(...handlers);
    }

    get(uri, ...handlers) {
        return this.add('GET', uri, ...handlers)
    }

    post(uri, ...handlers) {
        return this.add('POST', uri, ...handlers)
    }

    put(uri, ...handlers) {
        return this.add('PUT', uri, ...handlers)
    }

    patch(uri, ...handlers) {
        return this.add('PATCH', uri, ...handlers)
    }

    delete(uri, ...handlers) {
        return this.add('DELETE', uri, ...handlers)
    }

    /**
     * @name 添加路由
     * @param {String} method 
     * @param {String} uri 
     * @param  {...Handler} handlers 
     * @returns {router} 返回值为最后一级的router
     */
    add(method, uri, ...handlers) {
        if (uri.indexOf('/') == 0) uri = uri.substr(1);
        const uris = uri.split('/');
        let r = this;
        // 循环添加
        for(let u of uris) {
            let nr = new Router(u);
            // 判断
            if (u.indexOf(':') == 0) {
                nr.type = 'universal';
                nr.uri = nr.uri.substr(1);
                r.universalChildren.set(method, nr);
            } else if (u.indexOf('#') == 0) {
                nr.type = 'regex';
                nr.uri = nr.uri.substr(1);
                r.regexChildren.set(method, nr);
            } else {
                r.childrenRoute.set(`${method}-${u}`, nr);
            }
            r = nr;
        }
        // 添加路由处理及中间件
        r.middleware.push(...handlers);
        return r;
    }
    
    group(uri) {
        return this.add('ANY',uri)
    }
    
    prefix(uri) {
        return this.add('ANY',uri)
    }

    find(ctx) {
        let uris = [...ctx.url.substr(1).split('/'), ""];
        let r = this;
        ctx.middleware.push(...r.middleware || []);
        // 查询递归执行子路由
        for(let i = 0; i < uris.length; i++) {
            // 执行当前中间件
            let nr = r.childrenRoute.get(`${ctx.method}-${uris[i]}`) || r.childrenRoute.get(`ANY-${uris[i]}`) || null;
            if (r.regexChildren.get(ctx.method) && !nr) {
                nr = r.regexChildren.get(ctx.method) || null;
                if (!nr) break;

                let reg = new RegExp(nr.uri);
                if (!reg.test(uris[i])) {
                    // 正则匹配
                    let result = uris[i].match(reg);
                    if (result && result.groups) {
                        for(let k in result.groups) {
                            ctx.params.set(k, result.groups[k])
                        }
                    } else if (result) {
                        result.map((v,k) => {
                            ctx.params.set(`${i}-${k}`, v);
                        });
                    }
                }
            }

            if (r.universalChildren.get(ctx.method) && !nr) {
                nr = r.universalChildren.get(ctx.method) || null;
                if (nr) ctx.params.set(nr.uri, uris[i]);
            }
            // 如果没有查找到, 则查找模糊匹配项, 优先正则
            r = nr;
            if (r) {
                ctx.middleware.push(...r.middleware || []);
            } else {
                i = uris.length;
            }
        }

        // 如果没有查到最后的路由
        if (!r) ctx.middleware.push(async (ctx) => {
            ctx.statusCode = 404;
            ctx.send('Not found');
        });
        console.log(ctx)
        // 最后一次执行
        ctx.next();
    }
}

module.exports = Router;